### ISSUE TITLE
TorchScript possibly does not handle ternary conditional op correctly.

### ISSUE THREAD

**Issue by spandantiwari** [pytorch/pytorch#19047]
## 🐛 Bug

It seems like TorchScript does not handle ternary conditional op from Python (`[on_true] if [expression] else [on_false] `) correctly. 

## To Reproduce

1. The easiest way to repro this issue is by updating one of the existing test points.In the file torch/nn/modules/batchnorm.py change lines 62-65 from this:
```
if self.momentum is None:
        exponential_average_factor = 0.0
else:
        exponential_average_factor = self.momentum
```
to the following ternary op:
`exponential_average_factor = 0.0 if self.momentum is None else self.momentum`
2. Run test point in test/test_jit.py:
`python test_jit.py TestJitGeneratedModule.test_nn_BatchNorm2d_2d_simple_average`
3. The test will fail, even though functionally the two code pieces are functionally the same.

## Expected behavior

The above test point should pass.

## Environment

PyTorch version: 1.1.0a0+57e1255
Is debug build: Yes
CUDA used to build PyTorch: Could not collect

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.12.2

Python version: 3.6
Is CUDA available: No
CUDA runtime version: 7.5.17
GPU models and configuration: GPU 0: GeForce GTX 1080
Nvidia driver version: 384.130
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.2.1

Versions of relevant libraries:
[pip] numpy==1.15.1
[pip] torch==1.1.0a0+57e1255
[pip] torchvision==0.2.1
[conda] blas                      1.0                         mkl  
[conda] magma-cuda90              2.3.0                         1    pytorch
[conda] mkl                       2019.0                      118  
[conda] mkl-include               2019.0                      118  
[conda] mkl_fft                   1.0.4            py36h4414c95_1  
[conda] mkl_random                1.0.1            py36h4414c95_1  
[conda] torch                     1.1.0a0+57e1255           <pip>
[conda] torchvision               0.2.1                     <pip>

**Comment by spandantiwari** [comment]
cc @houseroad @suo

**Comment by eellison** [comment]
@wanchaol do you think this could be related to the `is None` compiling you did ?

**Comment by eellison** [comment]
There's a few things here:

1) We only do optional refinement on local variables, not attributes, so `self.momentum` won't get refined from `float?` to `float`. 

2) The reason it works without the ternary expression is because a sort of hack that was put in to allow us to compile `if is None` branches without first class modules.

3) Even if you assign a local with 
```
temp = self.momentum
exponential_average_factor = 0.0 if temp is None else temp
```
It doesn't currently compile. BatchNorm should be setting self.momentum to be type `float?`. Because it is currently being set to either float or None, the Optional refinement is not triggering. In the else branch instead of refining `float?` to `float` it is just returning a None value. This should be fixed with first class modules.

**Comment by wanchaol** [comment]
Ternary IF is handled different with general IF statement. We meta program IF var is None (this is kinda hack we did to only emit one branch when we are sure that the condition value is None), but if we change to ternary IF, we will only do optional refinement. But as @eellison said, there's no such a type refinement for module attributes yet.

**Comment by eellison** [comment]
docs are updated to say we don't refine module attributes and there is a separate smaller issue to track that now
